plugins/chrome-ux-report-backend/src/service/ChromeUXReportService.ts

```diff
@@ -0,0 +1,154 @@
+    try {
+      return this.database.getPeriodId(period);
+    } catch (error) {
+      this.logger.error(
+        `There is an error while getting period from database, error ${error.message}`,
+      );
+      throw error;
+    }
+  }
+
+  private async addOrigin(origin: string): Promise<boolean> {
+    try {
+      await this.database.addOrigin(origin);
+      return true;
+    } catch (error) {
+      this.logger.error(
+        `There is an error while adding origin to database, error ${error.message}`,
+      );
+      throw error;
+    }
+  }
+
+  private async addPeriod(period: string): Promise<boolean> {
+    try {
+      await this.database.addPeriod(period);
+      return true;
+    } catch (error) {
+      this.logger.error(
+        `There is an error while adding period to database, error ${error.message}`,
+      );
+      throw error;
+    }
+  }
+
+  private async addUXMetrics(
+    originId: number,
+    periodId: number,
+    metrics: Metric,
+  ): Promise<boolean> {
+    try {
+      await this.database.addUXMetrics({
+        origin_id: originId,
+        period_id: periodId,
+        fast_fp: metrics.fast_fp,
+        avg_fp: metrics.avg_fp,
+        slow_fp: metrics.slow_fp,
+        fast_fcp: metrics.fast_fcp,
+        avg_fcp: metrics.avg_fcp,
+        slow_fcp: metrics.slow_fcp,
+        fast_dcl: metrics.fast_dcl,
+        avg_dcl: metrics.avg_dcl,
+        slow_dcl: metrics.slow_dcl,
+        fast_ol: metrics.fast_ol,
+        avg_ol: metrics.avg_ol,
+        slow_ol: metrics.slow_ol,
+        fast_fid: metrics.fast_fid,
+        avg_fid: metrics.avg_fid,
+        slow_fid: metrics.slow_fid,
+        fast_ttfb: metrics.fast_ttfb,
+        avg_ttfb: metrics.avg_ttfb,
+        slow_ttfb: metrics.slow_ttfb,
+        small_cls: metrics.small_cls,
+        medium_cls: metrics.medium_cls,
+        large_cls: metrics.large_cls,
+        fast_lcp: metrics.fast_lcp,
+        avg_lcp: metrics.avg_lcp,
+        slow_lcp: metrics.slow_lcp,
+      });
+      return true;
+    } catch (error) {
+      this.logger.error(
+        `There is an error while getting ux metrics from database, error ${error.message}`,
+      );
+      return false;
+    }
+  }
+
+  async getUXMetrics(origin: string, period: string): Promise<Metric> {
+    try {
+      let originId = await this.getOriginId(origin);
+      let periodId = await this.getPeriodId(period);
+
+      if (originId && periodId) {
+        return this.database.getUXMetrics(originId, periodId);
+      }
+
+      if (!originId) {
+        await this.addOrigin(origin);
+        originId = await this.getOriginId(origin);
+      }
+
+      if (!periodId) {
+        await this.addPeriod(period);
+        periodId = await this.getPeriodId(period);
+      }
+
+      const rows = await this.queryClient.queryUXMetrics(origin, period);
+
+      await this.addUXMetrics(originId, periodId, rows);
+
+      return this.database.getUXMetrics(originId, periodId);
```

Comment:
If we already have the metrics in our hand, can't we return them instead of write + immediately read?

Just checking - we cache them once and they never go away or get updated? I presume this is because they are never changing once they are calculated right